compiler/rustc_borrowck/src/lib.rs

```diff
@@ -2282,28 +2283,29 @@ mod error {
         tainted_by_errors: Option<ErrorGuaranteed>,
     }
 
-    impl BorrowckErrors<'_> {
-        pub fn new() -> Self {
+    impl<'tcx> BorrowckErrors<'tcx> {
+        pub fn new(tcx: TyCtxt<'tcx>) -> Self {
             BorrowckErrors {
+                tcx,
                 buffered_move_errors: BTreeMap::new(),
                 buffered: Default::default(),
                 tainted_by_errors: None,
             }
         }
 
-        // FIXME(eddyb) this is a suboptimal API because `tainted_by_errors` is
-        // set before any emission actually happens (weakening the guarantee).
         pub fn buffer_error(&mut self, t: DiagnosticBuilder<'_, ErrorGuaranteed>) {
-            self.tainted_by_errors = Some(ErrorGuaranteed::unchecked_claim_error_was_emitted());
+            self.tainted_by_errors = Some(
```

Comment:
This only needs to set this once, instead of once per buffered error, right?

Reply:
`ErrorGuaranteed` is a ZST, so `Option<ErrorGuaranteed>` is "just" a typed `bool`.

Reply:
Yeah probably although its less code to just not check that and always do the assignment lol

Reply:
Yes, but it delays a bunch of extra bugs which make it difficult when you're using `-Ztreat-err-as-bug=N`, no?

Reply:
I really wish we would make treat-err-as-bug ignore delayed stuff and have a separate flag for that....